### PR TITLE
Implement DynamoDB Accelerator (DAX)

### LINKED ISSUE #250: Implement DynamoDb Accelerator (DAX)
AWS: DynamoDb Accelerator (DAX)

Amazon DynamoDB is designed for scale and performance. In most cases, the DynamoDB response times can be measured in single-digit milliseconds. However, there are certain use cases that require response times in microseconds. For these use cases, DynamoDB Accelerator (DAX) delivers fast response times for accessing eventually consistent data.
[more info](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DAX.html)

Console: [link](https://console.aws.amazon.com/dynamodb/home?region=us-east-1#cache-dashboard:)

API: [link](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/dax/DaxClient.html)

Following are other resources that needs to implemented
 - Cluster
 - Subnet groups
 - Parameter groups

Screen Shot 2020-05-15 at 4 27 48 PM

### PR DESCRIPTION
**Pull request by iHaoo**
Fixes #250

### CODE REVIEW

File: examples/dax/cluster.gyro [hunk 1]
```diff
@@ -0,0 +1,7 @@
+aws::dax-cluster cluster-example
+    name: "cluster-gyro-example"
+    iam-role-arn: "arn:aws:iam::242040583208:role/service-role/DAXtoDynamoDB"
```
Comment (deepanjan90) [inline examples/dax/cluster.gyro:3]: Create a role resource and associate that

File: examples/dax/subnet-group.gyro [hunk 2]
```diff
@@ -0,0 +1,5 @@
+aws::dax-subnet-group subnet-group
+    name: "subnet-group-example"
+    description: "subnet-group-description"
+    subnet-ids: ["subnet-07473edcb6aa2fff2"]
```
Comment (deepanjan90) [inline examples/dax/subnet-group.gyro:4]: Create a vpc and atleast 2 subnets and then associate those here as references

File: src/main/java/gyro/aws/dax/DaxClusterFinder.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *    dax-cluster: $(external-query aws::dax-cluster { names: "cluster-example"})
+ */
+@Type("dax-cluster")
+public class DaxClusterFinder extends AwsFinder<DaxClient, Cluster, DaxClusterResource> {
+
+    private List<String> names;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxClusterFinder.java:41]: The specific finder should only take in a single value for filtering

File: src/main/java/gyro/aws/dax/DaxClusterResource.java [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+    }
+
+    /**
+     * The list of availability zones of the cluster.
+     */
+    public List<String> getAvailabilityZones() {
+        if (availabilityZones == null) {
+            availabilityZones = new ArrayList<>();
+        }
+        return availabilityZones;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxClusterResource.java:221]:
```suggestion
        
        return availabilityZones;
```

File: src/main/java/gyro/aws/dax/DaxClusterResource.java [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+    /**
+     * The list of security group IDs assigned to each node in the cluster.
+     */
+    @Updatable
+    public List<String> getSecurityGroupIds() {
+        if (securityGroupIds == null) {
+            securityGroupIds = new ArrayList<>();
+        }
+        return securityGroupIds;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxClusterResource.java:318]:
```suggestion
        
        return securityGroupIds;
```

File: src/main/java/gyro/aws/dax/DaxNode.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    public String getAvailabilityZone() {
+        return availabilityZone;
+    }
+
+    public void setAvailabilityZone(String availabilityZone) {
+        this.availabilityZone = availabilityZone;
+    }
+
+    /**
+     * The endpoint of the node.
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxNode.java:44]:
```suggestion
     * The endpoint configuration of the node.
```

File: src/main/java/gyro/aws/dax/DaxNode.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            endpoint.copyFrom(model.endpoint());
+            setEndpoint(endpoint);
+        } else {
+            setEndpoint(null);
+        }
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("%s", getId());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxNode.java:121]: What of the id is null ?

File: src/main/java/gyro/aws/dax/DaxNode.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        setId(model.nodeId());
+        setStatus(model.nodeStatus());
+        setParameterGroupStatus(findById(DaxParameterGroupResource.class, model.parameterGroupStatus()));
+
+        if (model.endpoint() != null) {
+            DaxEndpoint endpoint = newSubresource(DaxEndpoint.class);
+            endpoint.copyFrom(model.endpoint());
+            setEndpoint(endpoint);
+        } else {
+            setEndpoint(null);
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxNode.java:115]: Set this before the if for consistency and code reduction

File: src/main/java/gyro/aws/dax/DaxParameterGroupFinder.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *    dax-parameter-group: $(external-query aws::dax-parameter-group { names: "parameter-group-example"})
+ */
+@Type("dax-parameter-group")
+public class DaxParameterGroupFinder extends AwsFinder<DaxClient, ParameterGroup, DaxParameterGroupResource> {
+
+    private List<String> names;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxParameterGroupFinder.java:41]: Should be singular as mentioned earlier.

File: src/main/java/gyro/aws/dax/DaxParameterGroupResource.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The list of name-value pairs for the parameters of the group.
+     */
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxParameterGroupResource.java:83]: Subresource tag missing.

File: src/main/java/gyro/aws/dax/DaxParameterGroupResource.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    @Override
+    public void create(GyroUI ui, State state) throws Exception {
+        DaxClient client = createClient(DaxClient.class);
+
+        client.createParameterGroup(r -> r
+            .description(getDescription())
+            .parameterGroupName(getName()));
+
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxParameterGroupResource.java:126]: No need to call this, as no output field present.

File: src/main/java/gyro/aws/dax/DaxParameterNameValue.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void copyFrom(ParameterNameValue model) {
+        setName(model.parameterName());
+        setValue(model.parameterValue());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("%s", getName());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxParameterNameValue.java:58]: What of name is null ?

File: src/main/java/gyro/aws/dax/DaxSecurityGroupMembership.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public void copyFrom(SecurityGroupMembership model) {
+        setSecurityGroupIdentifier(model.securityGroupIdentifier());
+        setStatus(model.status());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("%s", getSecurityGroupIdentifier());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxSecurityGroupMembership.java:58]: What if SecurityGroupIdentifier is null ?

File: src/main/java/gyro/aws/dax/DaxSubnet.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public void copyFrom(Subnet model) {
+        setAvailabilityZone(model.subnetAvailabilityZone());
+        setIdentifier(model.subnetIdentifier());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("%s", getIdentifier());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxSubnet.java:58]: What if Identifier is null ?

File: src/main/java/gyro/aws/dax/DaxSubnetGroupFinder.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *    dax-subnet-group: $(external-query aws::dax-subnet-group { names: "subnet-group-example"})
+ */
+@Type("dax-subnet-group")
+public class DaxSubnetGroupFinder extends AwsFinder<DaxClient, SubnetGroup, DaxSubnetGroupResource> {
+
+    private List<String> names;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxSubnetGroupFinder.java:41]: Should be singular. Same as earlier.

File: src/main/java/gyro/aws/dax/DaxSubnetGroupResource.java [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+ *        description: "subnet-group-description"
+ *        subnet-ids: ["subnet-07473edcb6aa2fff2"]
+ *    end
+ */
+@Type("dax-subnet-group")
+public class DaxSubnetGroupResource extends AwsResource implements Copyable<SubnetGroup> {
+
+    private String description;
+    private String name;
+    private List<String> subnetIds;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxSubnetGroupResource.java:56]: Should be a reference to the subnet resource

File: src/main/java/gyro/aws/dax/DaxSubnetGroupResource.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+    private List<DaxSubnet> subnets;
+    private String vpcId;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxSubnetGroupResource.java:58]: Should be a reference to the vpc resource

File: src/main/java/gyro/aws/dax/DaxTag.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    public void copyFrom(Tag model) {
+        setKey(model.key());
+        setValue(model.value());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("%s", getKey());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/dax/DaxTag.java:63]: What if key is null ?

**Review by deepanjan90 — CHANGES_REQUESTED** [review]
Inline changes.

keep mindful about two scenarios:
 - when associating a primary key
     - Check to see if the field can be null, if so, it cannot be used as a primary key
     - If only a single field is used for the primary key, string.format is not necessary.

File: src/main/java/gyro/aws/dax/DaxTag.java (hunk 18 above)
Comment (iHaoo) [reply src/main/java/gyro/aws/dax/DaxTag.java:63]: Doesn't @Required on key handle for a null key?

File: examples/dax/cluster.gyro [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+    replication-factor: 3
+    description: "test-description"
+end
+
+aws::iam-role iam-role-example
+    name: "dax-cluster-example-iam-role"
+    arn: "arn:aws:iam::242040583208:role/aws-service-role/dax.amazonaws.com/AWSServiceRoleForDAX"
+    assume-role-policy: "dax-role-policy.json"
+    policies: [$(external-query aws::iam-policy { arn: 'arn:aws:iam::aws:policy/AmazonECS_FullAccess' }).0]
+end
```
Comment (deepanjan90) [inline examples/dax/cluster.gyro:14]:
```suggestion
end

```

File: examples/dax/dax-role-policy.json [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "Service": "dax.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
```
Comment (deepanjan90) [inline examples/dax/dax-role-policy.json:12]:
```suggestion
}

```

**Pull request merged**